wdl-lint/src/rules/container_value.rs

```diff
@@ -0,0 +1,199 @@
+}
+
+/// Creates a diagnostic indicating that an array contains one or more 'any'
+/// URIs.
+fn array_containing_anys(spans: impl Iterator<Item = Span>) -> Diagnostic {
+    let mut diagnostic = Diagnostic::note(String::from("arrays containing any are ambiguous"))
+        .with_rule(ID)
+        .with_fix(format!(
+            "remove these entries or change the array to a string literal with the value of \
+             `{ANY_CONTAINER_VALUE}`"
+        ));
+
+    for span in spans {
+        diagnostic = diagnostic.with_highlight(span)
+    }
+
+    diagnostic
+}
+
+impl Rule for ContainerValue {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Ensures that values for the `container` key within `runtime`/`requirements` sections are \
+         well-formed."
+    }
+
+    fn explanation(&self) -> &'static str {
+        "This rule checks the following:
+
+        - Containers should have a tag, as container URIs with no tags have no expectation that \
+         the behavior of the containers won't change between runs.
+        - Further, immutable containers tagged with SHA256 sums are preferred. This is due to the \
+         requirement from the WDL specification that tasks produce functionally equivalent output \
+         across runs. When a mutable tag is used, there is a risk that changes to the container \
+         will cause different behavior between runs.
+        - Use of the 'any' container URI (`*`) within an array of container URIs is ambiguous and \
+         should be avoided.
+        - Empty container URI arrays are not disallowed by the specification but are ambiguous and \
+         should be avoided.
+        - An array of container URIs with a single element should be changed to a single string \
+         value."
+    }
+
+    fn tags(&self) -> TagSet {
+        // NOTE: these are the justification for these tags:
+        //
+        // - Clarity because it resolves the ambiguous situations described in the
+        //   explanation above.
+        // - Portability because this resolves situations where different execution
+        //   engines might behave differently (e.g., one container engine might always
+        //   pull the latest image for a mutably tagged container whereas another may
+        //   use a older, cached version until the user prompts it to upgrade).
+        TagSet::new(&[Tag::Clarity, Tag::Portability])
+    }
+}
+
+impl Visitor for ContainerValue {
+    type State = Diagnostics;
+
+    fn document(&mut self, _: &mut Self::State, _: VisitReason, _: &Document, _: SupportedVersion) {
+        // This callback is intentionally empty.
+    }
+
+    fn runtime_section(
+        &mut self,
+        state: &mut Self::State,
+        reason: VisitReason,
+        section: &RuntimeSection,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        if let Some(container) = section.container() {
```

Comment:
Should this lint warn if `container` is omitted?

Reply:
I think "yes". Although I could see implementing that as a separate lint? It should be warned on, either here or elsewhere.

Reply:
I tried this out in this rule and didn't like the affect. I think that all of the logic about which keys should be included/not included should be in the same rule as where we check to make sure certain sections are present _at all_. That way, we're not both flagging sections as being deprecated (such as a `runtime` block in a WDL v1.2 document) while also flagging within the runtime block that the `container` key is missing.

Reply:
This is tracked here: https://github.com/stjude-rust-labs/wdl/issues/153.